oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/banyandb/BanyanDBIndexInstaller.java

```diff
@@ -35,28 +36,32 @@
 @Slf4j
 public class BanyanDBIndexInstaller extends ModelInstaller {
     private final BanyanDBStorageConfig config;
+    private final ConfigService configService;
 
     public BanyanDBIndexInstaller(Client client, ModuleManager moduleManager, BanyanDBStorageConfig config) {
         super(client, moduleManager);
         this.config = config;
+        this.configService = moduleManager.find(CoreModule.NAME).provider().getService(ConfigService.class);
     }
 
     @Override
     protected boolean isExists(Model model) throws StorageException {
-        final MetadataRegistry.SchemaMetadata metadata = MetadataRegistry.INSTANCE.parseMetadata(model, config);
+        final MetadataRegistry.SchemaMetadata metadata = MetadataRegistry.INSTANCE.parseMetadata(model, config, configService);
         try {
             final BanyanDBClient c = ((BanyanDBStorageClient) this.client).client;
-            // first check group
-            Group g = metadata.getOrCreateGroup(c);
-            if (g == null) {
-                throw new StorageException("fail to create group " + metadata.getGroup());
+            // first check resource existence and create group if necessary
+            final boolean resourceExist = metadata.checkResourceExistence(c);
+            if (!resourceExist) {
+                return false;
             }
-            log.info("group {} created", g.name());
+
             // then check entity schema
             if (metadata.findRemoteSchema(c).isPresent()) {
-                MetadataRegistry.INSTANCE.registerModel(model, config);
+                MetadataRegistry.INSTANCE.registerModel(model, config, configService);
```

Comment:
I think we should keep existence checking here only. Otherwise, what is the point of `createTable`?

Reply:
Also, you have this 

```
// normally, we should not reach here
return false;
```

If it can't be reached, we should face an IllegalStateException

Reply:
Since I've removed this `registerModel` action here, it is now simplified

Reply:
This line cannot be removed. Since even if schema is already registered in the BanyanDB, we still have to register it locally for query and write.

So these lines are restored.

Reply:
Do we have a better to make codes more clear? Such as, when we do `isPresent`, could we use the metadata of existing measure/stream to build this local registration automatically?

Reply:
> Do we have a better to make codes more clear? Such as, when we do `isPresent`, could we use the metadata of existing measure/stream to build this local registration automatically?

I don't think it is enough to extract all necessary metadata with the schema in the BanyanDB. For example, 

```java
    public static class StreamToStorage implements Convert2Storage<StreamWrite> {
...

        @Override
        public void accept(String fieldName, Object fieldValue) {
            MetadataRegistry.ColumnSpec columnSpec = this.schema.getSpec(fieldName);
            if (columnSpec == null) {
                throw new IllegalArgumentException("fail to find field[" + fieldName + "]");
            }
            try {
                this.streamWrite.tag(fieldName, buildTag(fieldValue, columnSpec.getColumnClass()));
            } catch (BanyanDBException ex) {
                log.error("fail to add tag", ex);
            }
        }
    }

    private static Serializable<BanyandbModel.TagValue> buildTag(Object value, final Class<?> clazz) {
        if (int.class.equals(clazz) || Integer.class.equals(clazz)) {
            return TagAndValue.longTagValue(((Number) value).longValue());
        } else if (Long.class.equals(clazz) || long.class.equals(clazz)) {
            return TagAndValue.longTagValue((Long) value);
        } else if (String.class.equals(clazz)) {
            return TagAndValue.stringTagValue((String) value);
        } else if (Double.class.equals(clazz) || double.class.equals(clazz)) {
            return TagAndValue.binaryTagValue(ByteUtil.double2Bytes((double) value));
        } else if (StorageDataComplexObject.class.isAssignableFrom(clazz)) {
            return TagAndValue.stringTagValue(((StorageDataComplexObject<?>) value).toStorageData());
        } else if (clazz.isEnum()) {
            return TagAndValue.longTagValue((int) value);
        } else if (JsonObject.class.equals(clazz)) {
            return TagAndValue.stringTagValue((String) value);
        } else if (byte[].class.equals(clazz)) {
            return TagAndValue.stringTagValue((String) value);
        }
        throw new IllegalStateException(clazz.getSimpleName() + " is not supported");
    }
```

The mapping between data types in BanyanDB and OAP is not a surjection.

Reply:
I see, then how about we separating local registration logic as new methods?
Remote registration could call local registration automatically for sure.

Reply:
If we want to do so, I would suggest we add more overloading functions to the `Convert2Storage` interface. For example,

```java
public interface Convert2Storage<R> {
    void accept(String fieldName, String fieldValue);
    void accept(String fieldName, int fieldValue);
    void accept(String fieldName, long fieldValue);
    void accept(String fieldName, double fieldValue);
// etc.
}
```

One for each column type in OAP, so the type can be determined at compile time.

Reply:
> I see, then how about we separating local registration logic as new methods? Remote registration could call local registration automatically for sure.

`registerMeasureModel` and `registerStreamModel` are only responsible for registering local cache.

The `define` method is for remote registration.

Reply:
Got it. My fault. we should be good now.